### PR TITLE
Add cascading feature to propagate YAPIONData annotation to instance properties and their properties, too.

### PR DESCRIPTION
**Pull request by feueraustreter**
Hi!

Using YAPION in a legacy project makes it annoying to add the YAPIONData annotation to every class used in a complex and nested object tree.

Also there may be situations where it is not possible to change the source of used classes because they are part of an external library or because you do not have the source of these classes at all.

My idea is to add a flag to the YAPIONData annotation to just cascade the annotation to all classes used as properties / instance variables in an object and also to their children (properties / instance variables), too, and deeper.

During implementing this feature and using YAPION in one of my projects I found a problem in the enum code returning null in case of errors. I suggest using a specific exception instead. To be sure, everything works as before and to check the new exception I added some tests, too. Hope it's okay to include this change in this PR.

I am looking forward to your feedback
Torsten

### CODE REVIEW

**Comment by yoyosource** [comment]
Hello Torsten.

I appreciate your contribution and like you addition.
Thanks for adding some more test and there is no problem in including those changes in this PR.
I looked over you changes briefly this morning and will test them this week and will contact you after that.

Regards,
yoyosource

File: src/main/java/yapion/serializing/ContextManager.java [hunk 1]
```diff
@@ -100,11 +105,14 @@ YAPIONInfo is(Class<?> clazz) {
...
 
+    boolean isCascading() {
+        return yapionData != null && yapionData.cascading();
+    }
+
     YAPIONInfo is(Object object) {
         return is(object.getClass());
     }
 
-    @SuppressWarnings({"java:S3011"})
```
Comment (yoyosource) [inline src/main/java/yapion/serializing/ContextManager.java:107]: Why did you remove this Annotation.

File: src/main/java/yapion/serializing/YAPIONSerializer.java [hunk 2]
```diff
@@ -64,7 +64,11 @@ public YAPIONSerializer(@NonNull Object object, String context) {
 
     private YAPIONSerializer(@NonNull Object object, YAPIONSerializer yapionSerializer) {
         this.object = object;
-        this.contextManager = yapionSerializer.contextManager;
+        if (yapionSerializer.contextManager.isCascading()) {
+            this.contextManager = yapionSerializer.contextManager;
+        } else {
+            this.contextManager = new ContextManager(yapionSerializer.contextManager.get());
+        }
```
Comment (yoyosource) [inline src/main/java/yapion/serializing/YAPIONSerializer.java:71]: Is this for handling the YAPIONData.cascading property correctly?

File: src/main/java/yapion/serializing/ContextManager.java (hunk 1 above)
Comment (feueraustreter) [reply src/main/java/yapion/serializing/ContextManager.java]: I have deleted this by mistake

File: src/main/java/yapion/serializing/YAPIONSerializer.java (hunk 2 above)
Comment (feueraustreter) [reply src/main/java/yapion/serializing/YAPIONSerializer.java:71]: Basically, yes. 

If cascading is not used directly on the root object, but a little deeper in an object tree, the YAPIONData annotation with cascading should only be applied to deeper objects and not to parent objects. Therefore you cannot use the same context manager on all levels, sometimes you need a fresh one.

You will find the same code change in YAPIONDeserializer starting at line 154.

Comment (yoyosource) [reply src/main/java/yapion/serializing/YAPIONSerializer.java:71]: Thanks for the explanation.

File: src/main/java/yapion/serializing/ContextManager.java [hunk 3]
```diff
@@ -90,7 +90,12 @@ YAPIONInfo is(Class<?> clazz) {
         if (is(clazz.getDeclaredAnnotation(YAPIONSaveExclude.class))) globalSave = false;
         globalSave = is(clazz.getDeclaredAnnotation(YAPIONSave.class));
 
-        yapionData = clazz.getDeclaredAnnotation(YAPIONData.class);
+        if (yapionData == null) {
+            yapionData = clazz.getDeclaredAnnotation(YAPIONData.class);
+        }
+        if (!isCascading()) {
+            yapionData = clazz.getDeclaredAnnotation(YAPIONData.class);
+        }
```
[the remark this answers is not in the document]
Comment (yoyosource) [reply src/main/java/yapion/serializing/ContextManager.java:98]: I will look into making the cascading feature a more efficient one. in the serializer so that I do not need to create an Object each level until there is a cascading one. But that I will do myself.

[the remark this answers is not in the document]
Comment (feueraustreter) [reply src/main/java/yapion/serializing/ContextManager.java:98]: Great idea, I had not thought of that.

**Comment by yoyosource** [comment]
Thanks for your contribution. I would be pleased to see more pull request from you as well as anyone else.
I will merge this and close the pull request.

Regards,
yoyosource

File: src/main/java/yapion/serializing/ContextManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (yoyosource) [reply src/main/java/yapion/serializing/ContextManager.java:98]: Optimized with the commit: a87fe5bb4801b5c4259120cab61e26aed0c4211a
